### PR TITLE
[Core] Reduce unnecessary compute when logprobs=None

### PR DESCRIPTION
**Pull request by peng1999**
This PR refactors `_get_logprobs` function in sampler module,  to get rid of some unnecessary CPU and GPU works when `logprobs=None` in `SamplerParams`.

Tested using A100 80G GPU, running Qwen2-0.5B-Instruct and batch_size=2048. The latency of `_get_logprobs` reduced from ~30ms to ~5ms.

Before:

图片

After:

图片

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/sampler.py [hunk 1]
```diff
@@ -940,46 +950,53 @@ def _get_sampled_logprob_if_needed(
...
-                # num_logprobs + 1 for rank.
-                top_ranks = range(1, num_logprobs + 1)
-                sampled_logprobs_dict.update({
-                    top_id: (top_prob, rank)
-                    for top_id, top_prob, rank in zip(top_ids, top_probs,
-                                                      top_ranks)
+        if num_logprobs is None and not use_beam_search:
+            for next_token_id in next_token_ids:
+                # Use a dummy logprob
+                sampled_logprobs.append({next_token_id: Logprob(0.0)})
```
Comment (mgoin) [inline vllm/model_executor/layers/sampler.py:963]: Is 0 a value that makes sense for a logprob? We are still using real token ids here, so maybe 1.0 would be more representative?

Comment (peng1999) [reply vllm/model_executor/layers/sampler.py:963]: What do you think using NaN here? The value will propagate to `cumulative_logprobs`. I think it's better to show a NaN than an arbitrary positive number here.

**Review by rkooo567 — COMMENTED** [review]
QQ: is this breaking API change? What was the previous behavior when logprobs = None?

**Comment by peng1999** [comment]
> QQ: is this breaking API change? What was the previous behavior when logprobs = None?

The logprobs are already omitted if logprobs=None. So this API is not changing.

https://github.com/vllm-project/vllm/blob/1689219ebf01c750de492271832e27c39df38648/vllm/outputs.py#L124

The changed part is `cumulative_logprobs`. Previously it is calculated even if logprobs=None, but this PR makes it 0 (or NaN, see https://github.com/vllm-project/vllm/pull/6532#discussion_r1683031520) in this case. I think it's sensible. If users want to use cumulative_logprobs, they can just set logprobs=0.

**Review by comaniac — APPROVED** [review]
LGTM. Approve to unblock this PR.
Meanwhile, I agree that using NaN should be able to reduce some confusions.

File: vllm/model_executor/layers/sampler.py (hunk 1 above)
Comment (rkooo567) [reply vllm/model_executor/layers/sampler.py:963]: yeah + 1 in nan

**Review by rkooo567 — CHANGES_REQUESTED** [review]
can you add a simple test? and let's merge it after that!

**Review by mgoin — APPROVED** [review]
LGTM once NaN, thanks!

**Comment by peng1999** [comment]
It seems that some tests failed when I set the cumulative_logprobs to NaN because NaN != NaN. How should I deal with this?

**Comment by comaniac** [comment]
> It seems that some tests failed when I set the cumulative_logprobs to NaN because NaN != NaN. How should I deal with this?

It would be a bit tricky: https://stackoverflow.com/questions/13003202/python-nan-nan

I'd suggest one of the following:
1. Change the type of Logprob.logprob to `Optional[float]` and use `None` instead of `math.nan`. This is simpler, but I'm not sure if anywhere else in vLLM uses `logprobs` as float and may crash if the value is `None`.
2. Keep the type of Logprob.logprob as float, and assign 1.0 to it in this case. Additionally, add a new field `valid: bool` to indicate whether this value is valid or not. This makes Logprob fatter and more complicate, and it may still be mis-used if the consumer doesn't be aware of `valid`.

**Comment by rkooo567** [comment]
1 sounds like a good solution to me! can you actually update the sampling parameter docstring?

**Comment by peng1999** [comment]
I changed the dummy value from nan to inf for better comparability. The value will be discard eventually and end users will get both logprobs and cumulative_logprobs=None. I think the inf value is reasonable as an internal representation. Not using `None` inside `Logprob` to minimize code change.

The failed test seems to be irrelevant to this PR.

File: tests/samplers/test_logprobs.py [hunk 2]
```diff
@@ -63,7 +63,8 @@ def test_get_prompt_logprobs(
         assert result.outputs[0].logprobs is not None
         assert len(result.outputs[0].logprobs) == max_tokens
         for logprobs in result.outputs[0].logprobs:
-            assert len(logprobs) == num_top_logprobs
+            assert (len(logprobs) == num_top_logprobs
+                    or len(logprobs) == num_top_logprobs + 1)
```
Comment (comaniac) [inline tests/samplers/test_logprobs.py:67]: nit: Better to add a comment explaining the "+1" here as well.

**Review by comaniac — APPROVED** [review]
I'm ok with inf. Leave to @rkooo567 to approve and merge.

File: tests/samplers/test_logprobs.py (hunk 2 above)
Comment (peng1999) [reply tests/samplers/test_logprobs.py:69]: Addressed in fded571c610608e54080b741a63bff0c998bdea6.

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**